packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -281,16 +281,25 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
           return;
         }
 
-        const maybeDoc = this.serializer.fromDbRemoteDocument(dbRemoteDoc);
-        if (!query.path.isPrefixOf(maybeDoc.key.path)) {
-          control.done();
-        } else if (maybeDoc instanceof Document && query.matches(maybeDoc)) {
-          results = results.insert(maybeDoc.key, maybeDoc);
-        }
+        this.decodeDocAsync(dbRemoteDoc).next(maybeDoc => {
+          if (!query.path.isPrefixOf(maybeDoc.key.path)) {
+            control.done();
```

Comment:
In my opinion, IndexedDb is the worst API that was ever designed. Everything in IndexedDb is scoped by transactions, but the lifetime of a transaction essentially depends on the "threading behavior" of your code (or however one would call this on the Web). A transaction is committed to disk when the Javascript event loop returns control back to the browser and there are no outstanding IndexedDb operations. To put this in code:

```
indexedDb.openTransaction();
indexedDb.put(...);
indexedDb.put(...);
Promise.resolve().then(() => {
  // This put may fail if the transaction is no longer active. It may have been committed because 
  // Promise.resolve() is an asynchronous operation that momentarily returns control back from the 
  // event loop
  indexedDb.put(...);  
});
```

We work around this with the best API that was ever designed: Our PersistencePromise implementation is synchronous, but matches most other Promise semantics.

Long story short:
- What you have here works, but it doesn't parallelize document decoding since PersistencePromises block the event loop.
- If you change it to use a Promise, `control.done()` will be invoked after the transaction is committed.

Instead, you should be able to mirror what Android is doing:
- Decode the path first. If the path doesn't match, call `control.done()`.
- If the path matches, create a new Promise-returning arrow-function that decodes the document and insert it into the result set if the document matches. This function cannot access IndexedDb anymore.
- Before returning the results, use `Promise.all()` to wait for all results from the previous step.